### PR TITLE
Revert "kubeadm: Create control plane with ClusterFirstWithHostNet dns policy"

### LINKED ISSUE #1236: kubeadm: Control Plane with "ClusterFirstWithHostNet" is a circular dependency.
BUG REPORT 

## Versions

**kubeadm version** (use `kubeadm version`): 1.13

## What happened?
I was looking through the changelog for 1.13 and noticed:
https://github.com/kubernetes/kubernetes/pull/68890

With this change the apiserver will not be able to resolv dns names until kube-proxy and kube-dns are started
.
The resolv.conf of the apiserver when ClusterFirstWithHostNet is set:
```
bash-4.3# cat /etc/resolv.conf 
nameserver 10.96.0.10 #service ip of kube-dns
search default.svc.cluster.local svc.cluster.local cluster.local cisco.com
options ndots:5
```
This is likely green because the apiserver manifest has the following set:

```
--kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname
--etcd-servers=https://127.0.0.1:2379
```

With these settings we would probably make the tests green because kube-apiserver will be able to resolve what it needs to by using the ip addresses bypassing dns resolution.

If the user has an external etcd cluster this will break as the apiserver will be unable to resolve the external etcd cluster by name. 

If the kubelet-preferred-address-types is the default (Hostname) this will break as well. 

Any other calls the apiserver has to make before the kube-dns and kube-proxy are started will fail (Though there probably aren't too many calls to make there.)

### PR DESCRIPTION
**Pull request by neolit123**
Reverts kubernetes/kubernetes#68890

This is causing issues as described in the linked issue:
fixes kubernetes/kubeadm#1236

@kubernetes/sig-cluster-lifecycle-pr-reviews 
/priority critical-urgent
/kind bug
/assign @timothysc @chuckha 
/hold

### CODE REVIEW

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve 

/cc @andrewrynhard  - we needed to revert this!

**Comment by andrewrynhard** [comment]
@timothysc I see. Sorry for the troubles. I would still like to see `DNSClusterFirstWithHostNet` used but it sounds like its going to involve some more work.

**Comment by mauilion** [comment]
@andrewrynhard Since oidc url has to be https anyway in tectonic we did this by standing up an ingress controller and configuring that as part of our install. Being able to reference internal svc urls for this stuff is kind of brittle.

**Comment by andrewrynhard** [comment]
@mauilion Thanks for the suggestion. I actually ran it like that for quite some time. Using cert-manager and some rewrite rules in CoreDNS:
```
rewrite stop {
    name regex dex.example.com dex.security.svc.cluster.local
    answer name dex.security.svc.cluster.local dex.example.com
}
```
I am able to handle OIDC using only a service. It has worked out well so far. I can get valid certs for my internal service, but this whole setup is dependent on `DNSClusterFirstWithHostNet`.

**Comment by mauilion** [comment]
nice! this is a neat idea. This is a hard thing to solve since the apiserver comes in so early in the process. The solution to this is to ensure that the underlying node uses some dns server with dns stub capability or to ensure that before the apiserver starts a hosts entry like

10.96.0.100 dex.example.com
is added to the /etc/hosts file on the master nodes.
You *must* do this before the apiserver is started cause on start the apiserver will receive a copy of the /etc/hosts file of the node. 

TL;DR it's better to solve this problem without `ClusterFirstWithHostNet` because of the circular dependency.

Another possible and interesting solution would be to extend kubeadm to populate `hostAliases` on the control plane. 
This would enable you to set the /etc/hosts file for the apiserver for the endpoint directly I can see how this would be useful for a number of edge cases. 

as an example here is a modified /etc/kubernetes/manifests/kube-apiserver.yaml
```
~~ SNIP ~~
spec:
  hostAliases:
  - ip: "10.96.0.100"
    hostnames:
    - "dex.example.org"
  containers:
  - command:
    - kube-apiserver
    - --authorization-mode=Node,RBAC
    - --feature-gates=MountPropagation=true
~~ SNIP ~~
```
This results in a /etc/hosts on the apiserver pod that can resolv dex.example.org without resorting to `ClusterFirstWithHostNet`

**Comment by mauilion** [comment]
You can predefine it. You can specify the cluster ip when defining the Dex service. 
Example:

```
kubectl create service clusterip dex --tcp=3000 --clusterip=10.96.0.5  -l k8s-app=dex -n kube-system --dry-run -o yaml
```

**Comment by hansenwg** [comment]
@andrewrynhard Did you find a way to resolve this? I have a similiar scenario where I want to extend kube-apiserver by using the webhook with a service. However, DNS cannot resolve the `webhook.ns` and I could not fix an IP address due to potential conflicts

**Comment by andrewrynhard** [comment]
@hansenwg I kinda gave up on the idea unfortunately because I got pulled into other things, but I still would like to run it this way.

**Comment by hansenwg** [comment]
@andrewrynhard Thanks for the reply, I will look for some other workarounds then, thanks anyways!
